### PR TITLE
[Macros] Provide the Declared Type of Macros to the Constraint System

### PR DESCRIPTION
**Pull request by CodaFi**
Come with me on a magical journey of the layering of this patch.

We begin with a humble macro for `#line`, which has the declared interface type of `<T: ExpressibleByIntegerLiteral>` and returns said `T`. However, we cannot merely parse these signatures independently. Because there is no link between the two buffers, name lookup has very little hope of finding `T` (let alone the right `T`). So, instead, we invent a magical struct that binds them together - and thus the Macro Evaluation Context was born.

A macro evaluation context currently looks like this:

```
    struct __MacroEvaluationContext\(self.macro.genericSignature?.description ?? "") {
      typealias SignatureType = \(self.macro.signature)
    }
```

And so out pops a struct with the return type in context in a useful position for us. But we're not done yet. In order for this declaration to transit back to the compiler, it needs to run through the Swift parser, then be lowered through ASTGen, and finally passed back along the C interface and cached down in the AST context. All of this necessitated fixing up ASTGen to be able to handle generics, structs, and types in general.

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci test

File: lib/AST/ASTContext.cpp [hunk 1]
```diff
@@ -6037,3 +6040,17 @@ BuiltinTupleType *ASTContext::getBuiltinTupleType() {
 
   return result;
 }
+
+StructDecl *ASTContext::getOrCreateASTGenMacroContext(
```
Comment (DougGregor) [inline lib/AST/ASTContext.cpp:6044]: Why isn't this modeled as a request?

File: lib/Sema/CSGen.cpp [hunk 2]
```diff
@@ -1188,6 +1189,30 @@ namespace {
...
+        auto protocol =
+            TypeChecker::getLiteralProtocol(CS.getASTContext(), expr);
+        if (!protocol)
+          return Type();
+
+        auto openedType = CS.getTypeOfMacroReference(kind, expr);
+        if (!openedType)
+          return Type();
+
+        CS.addConstraint(ConstraintKind::LiteralConformsTo, openedType,
```
Comment (DougGregor) [inline lib/Sema/CSGen.cpp:1207]: Hmm, interesting. We don't have a way to express, in the surface language, that a particular conformance requirement is for a literal conformance. Maybe we can do something here so we don't need the hardcoded `getLiteralProtocol` and `LiteralConformsTo`.

File: lib/Sema/CSGen.cpp [hunk 3]
```diff
@@ -3603,7 +3628,17 @@ namespace {
     }
 
     Type visitMacroExpansionExpr(MacroExpansionExpr *expr) {
-      // FIXME: not implemented
+#if SWIFT_SWIFT_PARSER
+      auto &ctx = CS.getASTContext();
+      if (ctx.LangOpts.hasFeature(Feature::BuiltinMacros)) {
```
Comment (DougGregor) [inline lib/Sema/CSGen.cpp:3633]: I think this should be `Feature::Macros`.

File: lib/Sema/ConstraintSystem.cpp [hunk 4]
```diff
@@ -2469,6 +2470,27 @@ ConstraintSystem::getTypeOfMemberReference(
...
+
+  auto *locator = getConstraintLocator(anchor);
+  // Dig through to __MacroEvaluationContext.SignatureType
+  auto sig = getASTContext().getIdentifier("SignatureType");
+  auto *signature = macroCtx->lookupDirect(sig).front();
+  auto type = cast<TypeAliasDecl>(signature)->getUnderlyingType();
+
+  // Open up the generic type.
+  OpenedTypeMap replacements;
+  openGeneric(cast<TypeAliasDecl>(signature)->getParent(),
```
Comment (DougGregor) [inline lib/Sema/ConstraintSystem.cpp:2487]: Yeah, this makes perfect sense.

File: lib/Sema/TypeCheckMacros.cpp [hunk 5]
```diff
@@ -25,11 +25,71 @@
...
+        swift_ASTGen_getMacroEvaluationContext(
+            (const void *)start, (void *)(DeclContext *)macroSourceFile,
+            (void *)&ctx, macro, &context);
+
+        ctx.addCleanup([macro]() { swift_ASTGen_destroyMacro(macro); });
+        return dyn_cast<StructDecl>((Decl *)context);
+      });
+}
+
+#endif // SWIFT_SWIFT_PARSER
```
Comment (DougGregor) [inline lib/Sema/TypeCheckMacros.cpp:91]: I think this has to stay at the end of the file, so none of this code is compiled in when `SWIFT_SWIFT_PARSER` is not defined to `1`.

File: lib/Sema/TypeCheckMacros.h [hunk 6]
```diff
@@ -17,11 +17,40 @@
...
+/// }
+/// \endcode
+///
+/// So that we can use all of Swift's native name lookup and type resolution
+/// facilities to map the parsed signature type back into a semantic \c Type
+/// AST and a set of requiremnets.
+///
+/// \param macroName The name of the macro to look up.
+/// \param useDC The decl context of the use of this macro.
+StructDecl *lookup(StringRef macroName, DeclContext *useDC);
```
Comment (DougGregor) [inline lib/Sema/TypeCheckMacros.h:49]: Not important right now, but I think we should replace the `StructDecl *` with a `MacroDecl` or similar that captures the salient information for the macro and hides the struct-parsing hack a bit.

**Review by DougGregor — APPROVED** [review]
A couple of small comments, but this is great!

**Comment by CodaFi** [comment]
@swift-ci test

**Comment by CodaFi** [comment]
@swift-ci test Windows

**Comment by CodaFi** [comment]
@swift-ci smoke test

**Comment by CodaFi** [comment]
@swift-ci smoke test

**Pull request closed without merging**